lib/cylc/network/pyro_base.py

```diff
@@ -166,70 +145,77 @@ def _set_uri(self):
-                    uri_data["host"] = "localhost"
-                self.host = uri_data["host"]
+                    self.host = get_hostname()
 
         # Qualify the obj name with user and suite name (unnecessary but
         # can't change it until we break back-compat with older daemons).
         self.uri = (
-            'PYROLOC://%(host)s:%(port)s/%(owner)s.%(suite)s.%(target)s' %
-            uri_data)
-
-    def _get_proxy_common(self):
-        if self.pyro_proxy is None:
+            'PYROLOC://%(host)s:%(port)s/%(owner)s.%(suite)s.%(target)s' % {
+                "host": self.host,
+                "port": self.port,
+                "suite": self.suite,
+                "owner": self.owner,
+                "target": self.target_server_object})
+
+    def _get_proxy(self, reset=True, hash_name=None, cache_ok=True, old=False):
+        """Get a Pyro proxy."""
+        if reset or self.pyro_proxy is None:
             self._set_uri()
+            self.pphrase = self.reg_db.load_passphrase(
+                self.suite, self.owner, self.host, cache_ok)
             # Fails only for unknown hosts (no connection till RPC call).
             self.pyro_proxy = Pyro.core.getProxyForURI(self.uri)
             self.pyro_proxy._setTimeout(self.pyro_timeout)
-
-    def _get_proxy(self, hash_name=None):
-        self._get_proxy_common()
-        conn_val = ConnValidator()
-        if hash_name is not None and hash_name in OK_HASHES:
-            conn_val.set_default_hash(hash_name)
-        self.pyro_proxy._setNewConnectionValidator(conn_val)
-        self.pyro_proxy._setIdentification((self.my_uuid, self.pphrase))
-
-    def _get_proxy_old(self):
-        """Back compat: passphrase-only daemons (<= 6.4.1)."""
-        self._get_proxy_common()
-        self.pyro_proxy._setIdentification(self.pphrase)
+            if old:
+                self.pyro_proxy._setIdentification(self.pphrase)
+            else:
+                conn_val = ConnValidator()
+                if hash_name is None:
+                    hash_name = getattr(self, "_hash_name", None)
+                if hash_name is not None and hash_name in OK_HASHES:
+                    conn_val.set_default_hash(hash_name)
+                self.pyro_proxy._setNewConnectionValidator(conn_val)
+                self.pyro_proxy._setIdentification(
+                    (self.my_uuid, self.pphrase))
+        return self.pyro_proxy
 
     def reset(self):
+        """Reset pyro_proxy."""
         self.pyro_proxy = None
 
     def signout(self):
         """Multi-connect clients should call this on exit."""
         try:
-            self._get_proxy()
-            try:
-                self.pyro_proxy.signout()
-            except AttributeError:
-                # Back-compat for pre client reporting daemons <= 6.4.1.
-                pass
+            self._get_proxy().signout()
         except Exception:
             # Suite may have stopped before the client exits.
             pass
```

Comment:
Not intentionally, but I cannot see the reason why we need catch `AttributeError` separately here, now that `self._get_proxy()` returns `self.pyro_proxy`.